### DISCUSSION TITLE
Request for Comment: Web Development Tools

### DISCUSSION THREAD

**Discussion by morganmccunn** [Shopify/storefront-api-feedback#15] [RFC]
Today Shopify offers the [JS Buy SDK](http://shopify.dev/custom-storefronts/tools/js-buy) for merchants looking specifically to build websites powered by the Storefront API, and we're building [Hydrogen](http://shopify.dev/hydrogen) - Shopify’s recommended approach to building non-liquid custom web storefronts Hydrogen. 

This space is for any open feedback on web dev tooling with the Storefront API:
* What is working today?
* What would you expect Shopify to offer?
* What would make your life easier as a web developer building a custom website using the Storefront API?

Please let us know your feedback below 👇

**Comment by chrstrrs** [comment]
Is the Storefront API the same as the Shopify JS Buy SDK? Is the difference between the two is one JS and the other is GraphQL?

**Reply by johncraigcole** [reply]
Hey @wickedmom the JS Buy SDK is a lightweight Javascript library built on top of the Storefront API to help developers get up and running quickly. The library provides a wrapper around the underlying Storefront API GraphQL calls, and facilitates operations like product and collection queries and creating a checkout. 

If you want complete control over the API requests, you can make GraphQL requests to the Storefront API using a library of your choosing or even with native browser fetch calls.

**Reply by chrstrrs** [reply]
The JS Buy SDK is very easy and I personally love using it! Is the SDK still being worked on? It hasn't been updated for a year now. Some people are worried the project may be abandoned: https://github.com/Shopify/js-buy-sdk/issues/826

**Reply by morganmccunn** [reply]
Hey @wickedmom! Appreciate the feedback 🙏 We are looking understand how we support web developers moving forward and if we should be keeping the JS buy SDK as is, breaking it out to make it a less of a one-size-fits-all, or if the recently announced Hydrogen product will better satisfy wed dev needs. 

Can you elaborate a little bit more on your feedback? What was easy to use with it? Are you leveraging the pre-filled queries or building your own? What improvements (other than a version update!) would you love to see to make it even easier or more useful for you? Thanks for the input!

**Reply by lucadalli** [reply]
@morganmccunn As a user of the js-buy-sdk there's two things in particular which had a major negative impact on my DX:
1. Order of `lineItems` after an addition using `addLineItems` is non-deterministic.
2. Outdated/incorrect TypeScript types. (Given the surge of Typescript adoption, I think migrating the codebase from JS to TS would be worthwhile)

Also, are there plans to update the SDK to cover the new Storefront Cart API?

**Reply by morganmccunn** [reply]
Help, thanks @lucadalli. 

Also We have updated the latest SDK version here: https://github.com/Shopify/js-buy-sdk/releases/tag/v2.12.0

**Reply by gabrielvincent** [reply]
There's already a version `2022-07` release candidate and this major issue is still not fixed. Order of `lineItems` is still non-deterministic. How can this possibly be?
@morganmccunn Can we expect to have this fixed anytime soon?

**Comment by ahmed-adly-khalil** [comment]
Hello @morganmccunn 
Let's take https://www.netflix.shop as a great example to share some feedback assuming we would like to build a similar experience with Hydrogen.

### 1- content
- Content is one of the key areas that most custom storefronts struggle with and end up using 3p CMS, if we can avail CMS capabilities within Shopify that would make custom storefront builds much easier. Using meta fields could be an option. 
- Static site generation / re-generation: most of the speed we are seeing with headless commerce, comes from mixing static sites for primary page content with some dynamic functionality like shopping cart. having the ability to easily setup build/generation upon product/content updates would make a huge difference.

### 2- UX
- having a react library similar to app bridge but for storefronts where its components are able to automatically fetch store data and allow developers to set / customize templates would make building custom UX very easy
- Giving these components the ability to read from static site data would support build great UX without affecting performance 
- having a page builder at the backend of hydrogen that allows designers and developers to integrate custom react components in order to build pages like this: https://www.netflix.shop/pages/tracy-tubera-x-the-witcher 

### 3- apps / extensiblity 
- allowing Hydrogen to have `App extensions` is a great way to help people to build great experiences. most current headless solutions build all app functionality from scratch. 

### 4- billing
- allowing merchants to have the new hosting service oxygen as an add-on to their plan would make everybody's life easier.
